server/src/test/java/org/elasticsearch/action/ActionListenerTests.java

```diff
@@ -370,6 +372,53 @@ public void onFailure(Exception e) {
         assertThat(exReference.get(), instanceOf(IllegalArgumentException.class));
     }
 
+    public void testAssertAtLeastOnceWillLogAssertionErrorWhenNotResolved() throws Exception {
+        assumeTrue("assertAtLeastOnce will be a no-op when assertions are disabled", Assertions.ENABLED);
+        ActionListener<Object> listenerRef = ActionListener.assertAtLeastOnce(ActionListener.running(() -> {
+            // Do nothing, but don't use ActionListener.noop() as it'll never be garbage collected
+        }));
+        // Nullify reference so it becomes unreachable
+        listenerRef = null;
+        assertBusy(() -> {
+            System.gc();
+            assertLeakDetected("LEAK: resource was not cleaned up before it was garbage-collected\\.(.*|\\s)*");
+        });
+    }
+
+    public void testAssertAtLeastOnceWillNotLogWhenResolvedOrFailed() {
+        assumeTrue("assertAtLeastOnce will be a no-op when assertions are disabled", Assertions.ENABLED);
+        ReachabilityChecker reachabilityChecker = new ReachabilityChecker();
+        ActionListener<Object> listenerRef = reachabilityChecker.register(ActionListener.assertAtLeastOnce(ActionListener.running(() -> {
+            // Do nothing, but don't use ActionListener.noop() as it'll never be garbage collected
+        })));
+        // Call onResponse or onFailure
+        if (randomBoolean()) {
+            listenerRef.onResponse("succeeded");
+        } else {
+            listenerRef.onFailure(new RuntimeException("Failed"));
+        }
```

Comment:
I know an earlier iteration called the listener multiple times here and we said to remove that, but in fact I think we should call it multiple times to make sure another `assertOnce` doesn't creep back in. Sorry for the back-and-forth on this point :)

Reply:
Yep no worries, I nearly added that back after I made the change away from `runBefore`